src/libraries/System.Text.Json/tests/Common/ReferenceHandlerTests/ReferenceHandlerTests.IgnoreCycles.cs

```diff
@@ -429,6 +429,48 @@ public async Task IgnoreCycles_BoxedValueShouldNotBeIgnored()
             await Test_Serialize_And_SerializeAsync_Contains(root, expectedSubstring: @"""DayOfBirth"":15", expectedTimes: 2, s_optionsIgnoreCycles);
         }
 
+        [Fact]
+        public async Task IgnoreCycles_DerivedType_InArray()
+        {
+            var worker1 = new OfficeWorker
+            {
+                Office = new Office
+                {
+                    Dummy = new()
+                }
+            };
+
+            worker1.Office.Staff = [worker1, new RemoteWorker()];
+
+            await Test_Serialize_And_SerializeAsync(worker1, """{"Office":{"Staff":[null,{"$type":"remote"}],"Dummy":{}}}""", s_optionsIgnoreCycles);
+
+            worker1.Office.Staff = [worker1];
+
+            await Test_Serialize_And_SerializeAsync(worker1, """{"Office":{"Staff":[null],"Dummy":{}}}""", s_optionsIgnoreCycles);
```

Comment:
This looks good. Could you also include a test for the same type using `ReferenceHandler.Preserver`?

Reply:
IgnoreCycles and Preserve tests are a bit separated, so I just copied the classes over. I could otherwise just add another two lines in this test if youd like that better.